### PR TITLE
Protection

### PR DESCRIPTION
**Pull request by Dellybro**

### CODE REVIEW

File: contracts/Staking.sol [hunk 1]
```diff
@@ -1096,6 +1096,9 @@ contract Staking is IStaking, Initializable, AccessControlUpgradeable {
             uint256 shares,
             uint256 firstPayout
         ) = stakingV1.sessionDataOf(msg.sender, sessionId);
+
+        require(shares != 0, 'STAKING: Stake withdrawn v1');
```
Comment (Mvii7) [inline contracts/Staking.sol:1100]: .

File: contracts/Token.sol [hunk 2]
```diff
@@ -141,4 +158,88 @@ contract Token is
...
+            if (
+                toDisable == pair ||
+                toDisable == router ||
+                toDisable == address(0)
+            ) return; // Do nothing as we don't want to disable router
+
+            if (_timeOfLastTransfer[toDisable] == 0) {
+                _timeOfLastTransfer[toDisable] = block.timestamp;
+            } else {
+                require(
```
Comment (Mvii7) [inline contracts/Token.sol:185]: The idea is that if there are 2 transfers done in same block. With this logic the tx would simply fail. What we need to do is just don't do the 2nd transfer and let the tx through to persist state.

File: contracts/Token.sol [hunk 3]
```diff
@@ -141,4 +158,88 @@ contract Token is
...
+    function setPair(address _pair) external onlyManager {
+        pair = _pair;
+    }
+
+    function setRouter(address _router) external onlyManager {
+        router = _router;
+    }
+
+    function setTransferTime(uint256 _transferTime) external onlyManager {
+        transferTime = _transferTime;
```
Comment (Mvii7) [inline contracts/Token.sol:236]: what does this var do? is this like the time to wait between txs?

Comment (Mvii7) [reply contracts/Token.sol:236]: Should make the var name clearer

blocksBetweenTransfer or something like that

File: contracts/Auction.sol [hunk 4]
```diff
@@ -595,6 +595,25 @@ contract Auction is IAuction, Initializable, AccessControlUpgradeable {
         return auctionId;
     }
 
+    function removeTokensFromAuction(uint256 daysInFuture, uint256 amount)
```
Comment (Mvii7) [inline contracts/Auction.sol:598]: why is this here?

File: contracts/Token.sol [hunk 5]
```diff
@@ -11,15 +11,14 @@ import '@openzeppelin/contracts/token/ERC20/IERC20.sol';
...
-import 'hardhat/console.sol';
-
 contract Token is
     IToken,
     Initializable,
     ERC20Upgradeable,
     AccessControlUpgradeable
 {
     using SafeMathUpgradeable for uint256;
+    using EnumerableSetUpgradeable for EnumerableSetUpgradeable.AddressSet;
```
Comment (Mvii7) [inline contracts/Token.sol:21]: Not used

File: contracts/Token.sol [hunk 6]
```diff
@@ -141,4 +158,101 @@ contract Token is
...
+        virtual
+        override
+        isBlackedListed(_msgSender(), recipient)
+        returns (bool)
+    {
+        uint256 shouldTransfer = isTimeLimited(_msgSender(), recipient);
+        if (shouldTransfer == true) {
+            _transfer(_msgSender(), recipient, amount);
+        }
+        return true;
```
Comment (Mvii7) [inline contracts/Token.sol:226]: Return false

File: contracts/Token.sol [hunk 7]
```diff
@@ -141,4 +158,101 @@ contract Token is
...
+    function transfer(address recipient, uint256 amount)
+        public
+        virtual
+        override
+        isBlackedListed(_msgSender(), recipient)
+        returns (bool)
+    {
+        uint256 shouldTransfer = isTimeLimited(_msgSender(), recipient);
+        if (shouldTransfer == true) {
+            _transfer(_msgSender(), recipient, amount);
```
Comment (Mvii7) [inline contracts/Token.sol:224]: Return _transfer

File: contracts/Token.sol [hunk 8]
```diff
@@ -141,4 +158,101 @@ contract Token is
...
+     * - the caller must have a balance of at least `amount`.
+     */
+    function transfer(address recipient, uint256 amount)
+        public
+        virtual
+        override
+        isBlackedListed(_msgSender(), recipient)
+        returns (bool)
+    {
+        uint256 shouldTransfer = isTimeLimited(_msgSender(), recipient);
```
Comment (Mvii7) [inline contracts/Token.sol:222]: Don't need var for it

File: contracts/Token.sol [hunk 9]
```diff
@@ -141,4 +158,101 @@ contract Token is
...
+    function isTimeLimited(address sender, address recipient)
+        internal
+        returns (bool)
+    {
+        if (
+            timeLimited &&
+            notTimeLimited[recipient] == false &&
+            notTimeLimited[sender] == false
+        ) {
+            address toDisable;
```
Comment (Mvii7) [inline contracts/Token.sol:172]: toBlacklist makes more sense?

Comment (Dellybro) [reply contracts/Token.sol:172]: No we are "disabling" the mfrom selling for the transferTime allotted. If they happen to fall into the blacklist feature then they get blacklisted.

File: contracts/Token.sol (hunk 6 above)
Comment (Dellybro) [reply contracts/Token.sol:226]: No, we return true here, because either, the user is whitelisted, or transfertime is turned off. Therefore we want the transfer to succeed.

Comment (Mvii7) [reply contracts/Token.sol:225]: It will succeed, it will just tell the caller that transfer wasn't done
